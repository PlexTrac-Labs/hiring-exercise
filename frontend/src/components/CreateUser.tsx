import { ListItem, ListItemButton, ListItemIcon } from "@mui/material";
import React from "react";
import { LIST_MIN_WIDTH } from "../helpful/constants";
import AddCircleIcon from "@mui/icons-material/AddCircle";

export const CreateUser = () => {
  const style = {
    minWidth: LIST_MIN_WIDTH,
  };

  return (
    <ListItem sx={style}>
      <ListItemButton>
        <ListItemIcon>
          <AddCircleIcon color="success" />
        </ListItemIcon>
        Create User
      </ListItemButton>
    </ListItem>
  );
};
