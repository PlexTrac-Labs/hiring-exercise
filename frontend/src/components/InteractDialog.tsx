import { Dialog, DialogContent, DialogTitle } from "@mui/material";
import React from "react";
import { colors } from "../helpful/style";

export type InteractDialogProps = {
  open: boolean;
  title: string;
  handleClose: () => void;
};

export const InteractDialog: React.FC<InteractDialogProps> = (props) => {
  const { children, open, title, handleClose } = { ...props };

  const TitleStyle = {
    backgroundColor: colors.darkGreen,
    color: "white",
  };

  return (
    <Dialog open={open} onClose={handleClose}>
      <DialogTitle sx={TitleStyle}>{title}</DialogTitle>
      <DialogContent>{children}</DialogContent>
    </Dialog>
  );
};
