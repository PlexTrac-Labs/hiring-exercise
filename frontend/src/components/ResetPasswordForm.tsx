import { Alert, Button, FormControl, Grid, TextField } from "@mui/material";
import React from "react";

export type ResetFormProps = {
  error: boolean;
  errorMessage: string;
  handleSubmit: (event: any) => void;
};

const defaultProps: Partial<ResetFormProps> = {};

export const ResetPasswordForm = (props: ResetFormProps) => {
  const { error, errorMessage, handleSubmit } = {
    ...defaultProps,
    ...props,
  };
  return (
    <form onSubmit={handleSubmit}>
      <Grid
        container
        alignItems="center"
        justifyContent="center"
        rowSpacing={2}
        marginTop={3}
      >
        {error && (
          <Grid item xs={12}>
            <Alert severity="error">{errorMessage}</Alert>
          </Grid>
        )}
        <Grid item xs={12} sm={6}>
          <FormControl>
            <TextField
              required
              type="password"
              id="old"
              variant="outlined"
              label="Current Password"
            />
          </FormControl>
        </Grid>
        <Grid item xs={12} sm={6}>
          <FormControl>
            <TextField
              required
              type="password"
              id="new"
              variant="outlined"
              label="New Password"
            />
          </FormControl>
        </Grid>
        <Grid item xs={12} sm={6}>
          <FormControl>
            <TextField
              required
              type="password"
              id="confirm"
              variant="outlined"
              label="Confirm New Password"
            />
          </FormControl>
        </Grid>
        <Grid item xs={12}>
          <Button color="primary" variant="contained" type="submit">
            Change Password
          </Button>
        </Grid>
      </Grid>
    </form>
  );
};
