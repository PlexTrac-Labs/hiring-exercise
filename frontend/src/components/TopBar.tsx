import { AppBar, Button, Grid, Toolbar } from "@mui/material";
import { Box, styled } from "@mui/system";
import React from "react";
import { useNavigate } from "react-router-dom";
import { AUTH_TOKEN, TOP_BAR_HEIGHT } from "../helpful/constants";
import { colors } from "../helpful/style";

export const TopBar = () => {
  const loggedIn = !!localStorage.getItem(AUTH_TOKEN);
  const navigate = useNavigate();

  const Bar = styled(AppBar)(() => ({
    height: TOP_BAR_HEIGHT,
    backgroundColor: colors.darkGreen,
  }));

  const handleLogout = () => {
    localStorage.removeItem(AUTH_TOKEN);
    navigate("/");
  };

  return (
    <Bar>
      <Toolbar>
        {loggedIn && (
          <Grid
            container
            direction="row"
            justifyContent="space-between"
            alignItems="center"
          >
            <Grid item>
              <Button color="inherit" onClick={() => navigate("/list")}>
                Users
              </Button>
              <Button color="inherit" onClick={() => navigate("/reset")}>
                Reset Password
              </Button>
            </Grid>
            <Button color="inherit" onClick={handleLogout}>
              Logout
            </Button>
          </Grid>
        )}
      </Toolbar>
    </Bar>
  );
};
