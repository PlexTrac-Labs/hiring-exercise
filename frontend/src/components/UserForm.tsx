import {
  Alert,
  Button,
  DialogActions,
  FormControl,
  Grid,
  TextField,
} from "@mui/material";
import React from "react";
import { User } from "../util/user/types";

export type UserFormProps = {
  user?: User;
  showPassword?: boolean;
  submitText: string;
  error: boolean;
  errorMessage: string;
  handleSubmit: (event: any) => void;
  handleClose: () => void;
};

const defaultProps: Partial<UserFormProps> = {
  user: {
    _id: "",
    username: "",
    password: "",
    firstName: "",
    lastName: "",
    email: "",
    birthYear: 0,
    favColor: "",
    admin: false,
  },
  showPassword: false,
};

export const UserForm = (props: UserFormProps) => {
  const {
    user,
    showPassword,
    submitText,
    error,
    errorMessage,
    handleSubmit,
    handleClose,
  } = {
    ...defaultProps,
    ...props,
  };
  return (
    <form onSubmit={handleSubmit}>
      <Grid
        container
        alignItems="center"
        justifyContent="center"
        rowSpacing={2}
        marginTop={3}
      >
        {error && (
          <Grid item xs={12}>
            <Alert severity="error">{errorMessage}</Alert>
          </Grid>
        )}
        <Grid item xs={12} sm={6}>
          <FormControl>
            <TextField
              required
              id="username"
              variant="outlined"
              label="Username"
              defaultValue={user?.username}
            />
          </FormControl>
        </Grid>
        {showPassword && (
          <Grid item xs={12} sm={6}>
            <FormControl>
              <TextField
                required
                type="password"
                id="password"
                variant="outlined"
                label="Password"
                defaultValue={user?.username}
              />
            </FormControl>
          </Grid>
        )}
        <Grid item xs={12} sm={6}>
          <FormControl>
            <TextField
              required
              id="firstName"
              variant="outlined"
              label="First Name"
              defaultValue={user?.firstName}
            />
          </FormControl>
        </Grid>
        <Grid item xs={12} sm={6}>
          <FormControl>
            <TextField
              required
              id="lastName"
              variant="outlined"
              label="Last Name"
              defaultValue={user?.lastName}
            />
          </FormControl>
        </Grid>
        <Grid item xs={12} sm={6}>
          <FormControl>
            <TextField
              required
              id="email"
              variant="outlined"
              label="Email"
              defaultValue={user?.email}
            />
          </FormControl>
        </Grid>
        <Grid item xs={12} sm={6}>
          <FormControl>
            <TextField
              required
              type="number"
              id="birthYear"
              variant="outlined"
              label="Birth Year"
              defaultValue={user?.birthYear === 0 ? 2000 : user?.birthYear}
            />
          </FormControl>
        </Grid>
        <Grid item xs={12} sm={6}>
          <FormControl>
            <TextField
              required
              id="favColor"
              variant="outlined"
              label="Favorite Color"
              defaultValue={user?.favColor}
            />
          </FormControl>
        </Grid>
      </Grid>
      <DialogActions>
        <Button color="success" variant="contained" type="submit">
          {submitText}
        </Button>
        <Button onClick={handleClose}>Cancel</Button>
      </DialogActions>
    </form>
  );
};
