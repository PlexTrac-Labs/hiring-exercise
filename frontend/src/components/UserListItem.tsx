import DeleteIcon from "@mui/icons-material/Delete";
import {
  IconButton,
  ListItem,
  ListItemButton,
  ListItemText,
  Tooltip,
} from "@mui/material";
import React from "react";
import { LIST_MIN_WIDTH } from "../helpful/constants";
import { User } from "../util/user/types";

export type UserListItemProps = {
  user: User;
  handleDelete: () => void;
  handleClick: (user: User) => void;
};

export const UserListItem = (props: UserListItemProps) => {
  const { user, handleDelete, handleClick } = { ...props };

  return (
    <ListItem
      sx={{ minWidth: LIST_MIN_WIDTH }}
      secondaryAction={
        <Tooltip title="Delete User" placement="right" enterDelay={1000}>
          <IconButton edge="end" onClick={handleDelete}>
            <DeleteIcon />
          </IconButton>
        </Tooltip>
      }
    >
      <ListItemButton onClick={() => handleClick(user)}>
        <ListItemText>{user.username}</ListItemText>
      </ListItemButton>
    </ListItem>
  );
};
