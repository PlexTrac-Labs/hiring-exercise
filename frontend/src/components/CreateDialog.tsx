import React, { useState } from "react";
import { CreateUserRequest, User } from "../util/user/types";
import { createUser, getAllUsers } from "../util/user/user";
import { InteractDialog } from "./InteractDialog";
import { UserForm } from "./UserForm";

export type CreateDialogProps = {
  open: boolean;
  setUsers: (users: User[]) => void;
  onClose: () => void;
};

export const CreateDialog = (props: CreateDialogProps) => {
  const { open, setUsers, onClose } = { ...props };
  const [error, setError] = useState(false);
  const errorMessage = "Problem creating user. Is this email already in use?";

  const handleClose = () => {
    setError(false);
    onClose();
  };

  const handleCreate = async (event: any) => {
    event.preventDefault();
    const payload: CreateUserRequest = {
      username: event.target.username.value,
      password: event.target.password.value,
      firstName: event.target.firstName.value,
      lastName: event.target.lastName.value,
      email: event.target.email.value,
      birthYear: parseInt(event.target.birthYear.value),
      favColor: event.target.favColor.value,
    };
    try {
      await createUser(payload);
      let users = await getAllUsers();
      setUsers([...users]);
      handleClose();
    } catch (error: any) {
      setError(true);
    }
  };

  return (
    <InteractDialog open={open} handleClose={handleClose} title="Create User">
      <UserForm
        showPassword
        submitText="Create"
        error={error}
        errorMessage={errorMessage}
        handleSubmit={handleCreate}
        handleClose={handleClose}
      />
    </InteractDialog>
  );
};
