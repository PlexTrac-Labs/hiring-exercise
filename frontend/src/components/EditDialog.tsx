import React, { useState } from "react";
import { UpdateUserRequest, User } from "../util/user/types";
import { getAllUsers, updateUser } from "../util/user/user";
import { InteractDialog } from "./InteractDialog";
import { UserForm } from "./UserForm";

export type EditDialogProps = {
  open: boolean;
  user: User;
  setUsers: (users: User[]) => void;
  onClose: () => void;
};

export const EditDialog = (props: EditDialogProps) => {
  const { open, user, setUsers, onClose } = { ...props };
  const [error, setError] = useState(false);
  const errorMessage = "Problem updating user. Is this email already in use?";

  const handleClose = () => {
    setError(false);
    onClose();
  };

  const handleUpdate = async (event: any) => {
    event.preventDefault();
    const payload: UpdateUserRequest = {
      username: event.target.username.value,
      firstName: event.target.firstName.value,
      lastName: event.target.lastName.value,
      email: event.target.email.value,
      birthYear: parseInt(event.target.birthYear.value),
      favColor: event.target.favColor.value,
    };
    try {
      await updateUser(user._id, payload);
      let users = await getAllUsers();
      setUsers([...users]);
      handleClose();
    } catch (error: any) {
      setError(true);
    }
  };

  return (
    <InteractDialog open={open} handleClose={handleClose} title="Edit User">
      <UserForm
        user={user}
        submitText="Save"
        error={error}
        errorMessage={errorMessage}
        handleSubmit={handleUpdate}
        handleClose={handleClose}
      />
    </InteractDialog>
  );
};
