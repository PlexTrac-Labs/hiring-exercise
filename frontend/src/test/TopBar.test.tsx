import { cleanup, render } from "@testing-library/react";
import React from "react";
import { TopBar } from "../components/TopBar";

const mockedUsedNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...(jest.requireActual("react-router-dom") as any),
  useNavigate: () => mockedUsedNavigate,
}));

afterEach(cleanup);

it("renders", async () => {
  const { asFragment } = render(<TopBar />);
  expect(asFragment()).toMatchSnapshot();
});
