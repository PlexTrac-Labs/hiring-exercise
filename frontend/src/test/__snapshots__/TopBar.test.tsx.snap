// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<DocumentFragment>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-hntkvi-MuiPaper-root-MuiAppBar-root"
    data-testid="topbar"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1t29gy6-MuiToolbar-root"
    />
  </header>
</DocumentFragment>
`;
